Limit tree depth when blending Calculation tree with Blend node

anttijk opened the pull request:
#### 63f824e54433a84d58d6186d9e4dc27a66b4f7eb
<pre>
Limit tree depth when blending Calculation tree with Blend node
<a href="https://bugs.webkit.org/show_bug.cgi?id=302401">https://bugs.webkit.org/show_bug.cgi?id=302401</a>
<a href="https://rdar.apple.com/164138229">rdar://164138229</a>

Reviewed by Sam Weinig.

When blending calc() values we in complex cases add Blend(treeA, treeB, percent) nodes
to the calculation tree. The values being blended may themselves have been output of blending
and contain Blend nodes already. It is possible for the nodes to accumulate creating
a very deep tree. Recursive algorithms on such tree may blow up the stack.

* Source/WebCore/style/calc/StyleCalculationTree+Traversal.h:
(WebCore::Style::Calculation::forAllChildren):

Fix this to work in general case.

* Source/WebCore/style/calc/StyleCalculationTree.cpp:
(WebCore::Style::Calculation::computeDepth):

Add a function for computing the tree depth.

* Source/WebCore/style/calc/StyleCalculationTree.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper+Blending.h:
(WebCore::Style::LengthWrapperBlendingSupport::blendMixedSpecifiedTypes):

If the tree depth is already &gt;128 don&apos;t create a Blend node and just return the target tree.

Canonical link: <a href="https://commits.webkit.org/302921@main">https://commits.webkit.org/302921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a3d77bc595e9c15a98e219aed57ab0b0c14a5cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138069 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82273 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4337b0a6-e879-4ce2-aed1-f56949280959) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132519 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2812 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99535 "Found 1 new test failure: imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a1ea9256-71ab-4619-8229-c5f02e8bcefd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133595 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/116984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80243 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/db386beb-2525-4811-812e-33c5abdea1dd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35117 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81325 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110628 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140549 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2709 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/2453 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108040 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107972 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/27474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2094 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31754 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55697 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2779 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66168 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2598 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2800 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2705 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig reviewed (approved):
I think we need the same fix in `Blending<LengthPercentage<R, V>>`, so maybe moving the Calculation::blend() function out of line (and into StyleCalculationTree.cpp) and putting the depth guard there would make more sense.

I actually think we can make this case work without nesting blend nodes, but having this in place first makes sense.

anttijk committed:
Limit tree depth when blending Calculation tree with Blend node
https://bugs.webkit.org/show_bug.cgi?id=302401
rdar://164138229

Reviewed by Sam Weinig.

When blending calc() values we in complex cases add Blend(treeA, treeB, percent) nodes
to the calculation tree. The values being blended may themselves have been output of blending
and contain Blend nodes already. It is possible for the nodes to accumulate creating
a very deep tree. Recursive algorithms on such tree may blow up the stack.

* Source/WebCore/style/calc/StyleCalculationTree+Traversal.h:
(WebCore::Style::Calculation::forAllChildren):

Fix this to work in general case.

* Source/WebCore/style/calc/StyleCalculationTree.cpp:
(WebCore::Style::Calculation::computeDepth):

Add a function for computing the tree depth.

* Source/WebCore/style/calc/StyleCalculationTree.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper+Blending.h:
(WebCore::Style::LengthWrapperBlendingSupport::blendMixedSpecifiedTypes):

If the tree depth is already >128 don't create a Blend node and just return the target tree.

Canonical link: https://commits.webkit.org/302921@main
